Add `ql:has-word` triples if option `--add-has-word-triples`

hannahbast opened the pull request:
Optionally add, for each triple with a literal object, and for each word in that literal, an internal triple `<literal> ql:has-word "word"`. The number of occurrences of the word in the literal is stored in the graph . This can be activated by calling `IndexBuilderMain` with option `--add-has-word-triples`, or by setting `ADD_HAS_WORD_TRIPLES = true` in the `[index]` section of the `Qleverfile`. When embedded into C++ code, activate using `config.addHasWordTriples_ = true`.

These triples can be used for customized text search. To make this efficient, materialized views can be used. For example, to enable full-text search in the `rdfs:label` and `skos:altLabel` literals of all subjects in your dataset, giving weight 5 to the former and weight 2 to the latter, create a materialized view (using `qlever materialized-view words "..."`) from this query:
```sparql
SELECT ?word ?subject ?score WHERE {
  { ?subject rdfs:label ?text BIND (5 AS ?weight) } UNION { ?subject skos:altLabel ?text BIND (2 AS ?weight) }
  GRAPH ?tf { ?text ql:has-word ?word }
  BIND (?tf * ?weight AS ?score)
}
```
This view can then be efficiently queried as follows
```sparql
SELECT ?subject (?s1 + ?s2 + ?s3 AS ?score) WHERE {
  SERVICE view:words { [ view:column-word "keyword1" ; view:column-subject ?subject; view:column-score ?s1 ] }
  SERVICE view:words { [ view:column-word "keyword2" ; view:column-subject ?subject; view:column-score ?s2 ] }
  SERVICE view:words { [ view:column-word "keyword3" ; view:column-subject ?subject; view:column-score ?s3 ] }
}
```

NOTE 1: On the side, improve variable names and comments surrounding the changes made here. There is more work to do in this respect, which should be done in a separate PR. See the various TODOs added.

NOTE 2: The `ql:has-word` triples currently remain unchanged when there are update requests. This, too, is work for a separate PR

Commit:
Add `ql:has-word` triples to internal PSO&POS permutation

During parsing, for each triples with a literal object, and for each
word in that literal, add an internal triple `subject ql:has-word
"word"`.

TODO: This is currently done unconditionally, which makes it easier to
test (we don't need special options in the Qleverfile). Eventually,
there should be an option `--add-has-word-triples` to `IndexBuilderMain`
to enable this behavior. Tests are also still missing

Commit:
The subject of the `ql:has-word` triple should be the literal

Commit:
Add option `--add-has-word-triples` to `IndexBuilderMain`

Commit:
Fix failing unit tests

Commit:
Add word positions (as graph) and log number of triples added

Commit:
Write term frequency, not position

Writing the position is more general. But computing the term frequencies
for each text-word pair is currently not efficient in QLever (it
requires too much memory and a GROUP BY with two variables is much
slower than a GROUP BY with one variable). Since we never needed
positions so far, but we do want term frequencies for scoring, let's
make this the default for now.

Commit:
Cleaned up the code and many comments (also in related code)

Commit:
Merge remote-tracking branch 'origin/master' into add-text-triples

RobinTF reviewed (commented):
Mainly a lot of comments

Commit:
Merge remote-tracking branch 'origin/master' into add-text-triples

Commit:
Merge remote-tracking branch 'origin/master' into add-text-triples

Commit:
Propagate settings for handling integer overflow to all RDF parsers

So far, `RdfParallelParser`,  `RdfStreamParser`, and `RdfStringParser`
never received the settings for `integerOverflowBehavior` and
`invalidLiteralsAreSkipped`. Now they do.

Commit:
Merge remote-tracking branch 'origin/master' into add-text-triples

hannahbast commented on src/index/IndexBuilderTypes.h:
```diff
@@ -54,29 +54,47 @@ struct TripleComponentWithIndex {
+    serializer | arg.iriOrLiteral_;
+    serializer | arg.isExternal_;
+  }
+};
```
I found the order more meaningful that way. A lot of the changes in this file are better names for poorly named variables and (much) better comments.

hannahbast commented on src/index/IndexBuilderTypes.h:
```diff
@@ -239,87 +261,120 @@ struct LangtagAndTriple {
+    // In each map, assign the first IDs to the special IRIs `ql:langtag` and
+    // `ql:has-word`.
+    //
+    // NOTE: This is not necessary for functionality, but certain unit tests
```
I turned the TODO into a NOTE, but didn't remove anything

hannahbast commented on src/index/IndexBuilderTypes.h:
```diff
@@ -239,87 +261,120 @@ struct LangtagAndTriple {
+      //
+      // NOTE: There is similar code in `DeltaTriples::makeInternalTriples`
+      // for adding these internal triples for update triples. If you change
+      // this code, you probably also have to change that one.
```
It would be great to make this work with updates. I consider this work for a separate PR. And it would be awesome if you could do it. I think you know best what to do here

hannahbast commented on src/index/IndexBuilderTypes.h:
```diff
@@ -239,87 +261,120 @@ struct LangtagAndTriple {
+      // Third, if applicable, add a `ql:has-word` triple for each distinct word
+      // in the literal. We abuse the graph ID field to store the term
+      // frequency of the word in the literal.
+      if (!lt.wordFrequencies_.empty()) {
```
I already did a lot of renaming and minor refactoring on the side. I think a bigger refactoring is outside the scope of this PR

Commit:
Set `addHasWordTriples_ = false` by default in `Qlever.h`

And add a few more TODOs and expand some NOTEs

Commit:
Revert unrelated change regarding `integerOverflowBehavior` and `invalidLiteralsAreSkipped`

RobinTF commented on src/index/IndexBuilderTypes.h:
```diff
@@ -132,14 +150,20 @@ struct ItemMapAndBuffer {
+  ItemMapAndBuffer map_;
+  ad_utility::HashMap<Id, Id> specialIdMapping_;
+  uint64_t minId_ = 0;
+  const TripleComponentComparator* comparator_ = nullptr;
```
There's no reason to default initialise these, as the constructor always overwrites these values.

hannahbast replied:
Done

RobinTF commented on src/index/IndexBuilderTypes.h:
```diff
@@ -239,87 +261,129 @@ struct LangtagAndTriple {
+    if (index->addHasWordTriples()) {
+      itemMaps[j]->getId(TripleComponent{
+          ad_utility::triple_component::Iri::fromIriref(HAS_WORD_PREDICATE)});
+    }
```
This should be completely redundant. In #2839 I remove the line above this. `HAS_WORD_PREDICATE` will always get added automatically, even without this code. No need to add anything.

hannahbast replied:
Done

RobinTF commented on src/index/IndexBuilderTypes.h:
```diff
@@ -239,87 +261,129 @@ struct LangtagAndTriple {
+      // We return processed versions of: (1) the original triple, (2) two
+      // internal triples for the language tag (if any), and (3) one triple for
+      // each distinct word in the literal (if applicable).
+      IdTriples result(3 + lt.wordFrequencies_.size());
```
Isn't this `std::vector` super expensive in comparison to `std::array`? This lambda is called very often, and for each call memory is allocated on the heap. Consider using `absl::InlinedVector<IdTriple, 3>` to avoid this allocation for most cases. (Also please note the subtle hint that I believe the optional is probably unnecessary here.)

hannahbast replied:
Good catch + done

RobinTF commented on src/index/IndexBuilderTypes.h:
```diff
@@ -239,87 +261,129 @@ struct LangtagAndTriple {
+          // Add the internal triple `<literal> ql:has-word "word"`.
+          auto wordId = map.getId(TripleComponent{
+              ad_utility::triple_component::Literal::fromEscapedRdfLiteral(
+                  absl::StrCat("\"", word, "\""))});
```
I have to agree with the copilot here, I don't think there's any reason to use `fromEscapedRdfLiteral` here, using `literalWithoutQuotes` also is more efficient.

hannahbast replied:
Done

RobinTF commented on src/index/IndexBuilderTypes.h:
```diff
@@ -239,87 +261,120 @@ struct LangtagAndTriple {
+        // Update the counter for the number of ql:has-word triples.
+        if (numHasWordTriples != nullptr) {
+          numHasWordTriples->fetch_add(lt.wordFrequencies_.size(),
+                                       std::memory_order_relaxed);
```
Also bumping this comment. Whenever `std::memory_order_relaxed` is used it should come with a comment nearby that explains why this is absolutely necessary.

hannahbast replied:
Your are right + I added a comment

RobinTF commented on src/index/IndexBuilderTypes.h:
```diff
@@ -239,87 +261,129 @@ struct LangtagAndTriple {
-  // setup a tuple with one lambda function per map in the itemArray
-  // (the first lambda will assign ids according to itemArray[1]...
+  // Return one of the above lambdas for each thread.
   auto itemMapLambdaTuple =
       ad_tuple_helpers::setupTupleFromCallable<NumThreads>(itemMapLamdaCreator);
   return itemMapLambdaTuple;
```
Consider combining these lines.

hannahbast replied:
Done

RobinTF commented on src/index/IndexImpl.cpp:
```diff
@@ -513,6 +515,8 @@ IndexBuilderDataAsExternalVector IndexImpl::passFileForVocabulary(
   ad_utility::CachingMemoryResource cachingMemoryResource;
   ItemAlloc itemAlloc(&cachingMemoryResource);
+  // Counter for the number of ql:has-word triples created.
+  std::atomic<size_t> numHasWordTriples = 0;
```
This design choice goes against the general design choice of `getIdMapLambdas`, which explicitly creates a datastructure for each thread so that all of them can operate in parallel, but I can see why it is convenient to do this.

hannahbast replied:
Yes, it's convenient and I don't see that it impacts performance negatively, so will leave it as is

RobinTF reviewed (changes requested):
Mostly minor suggestions.

Commit:
Address all comments from Robin's review

In particular, now use `absl::InlinedVector<IdTriple, 3>`. The old code
used a `std::array<std::optional<IdTriple>, 3>`, the last previous
commit used `std::vector<std::optional<IdTriple>>`. We now avoid the
heap allocation for the common case of 3 triples, as well as the
`std::optional`

RobinTF commented on src/index/IndexImpl.cpp:
```diff
@@ -543,11 +543,9 @@ IndexBuilderDataAsExternalVector IndexImpl::passFileForVocabulary(
+        for (const auto& triple : opt.value()) {
+          actualCurrentPartialSize++;
+          localWriter.push_back(triple);
         }
```
```suggestion
        const auto& triples = opt.value();
        actualCurrentPartialSize += triples.size();
        localWriter.insert(localWriter.end(), triples.begin(), triples.end());
```

This should in theory allow the compiler to more efficiently copy the data.

Commit:
Addressed the last two comments as well

RobinTF reviewed (approved):
Thank you very much

Commit:
Default `IndexImpl::addHasWordTriples_` to `false`

This didn't really have an effect because the value is overwritten by
whatever comes from the command-line option `--add-has-word-triples`
(which defaults to `false`), but it makes it more clear that the default
value is `false` and not `true`.

Commit:
Merge remote-tracking branch 'origin/master' into add-text-triples

Commit:
Merge remote-tracking branch 'origin/master' into add-text-triples

Commit:
Make clang-format happy
